tools/clang/unittests/HLSLExec/LongVectors.h

```diff
@@ -418,101 +545,99 @@ template <typename DataTypeT> class TestConfig {
+                                                   L"DefaultInputValueSet3"};
 
 protected:
   // Prevent instances of TestConfig from being created directly. Want to force
   // a derived class to be used for creation.
   template <typename OpTypeT>
   TestConfig(const OpTypeMetaData<OpTypeT> &OpTypeMd)
       : OpTypeName(OpTypeMd.OpTypeString), Intrinsic(OpTypeMd.Intrinsic),
-        Operator(OpTypeMd.Operator) {}
-
-  // The appropriate computeExpectedValue should be implemented in derived
-  // classes. Impelemented as virtual here to prevent requiring all derived
-  // classes from needing to implement. The OS builds disable RTTI, so using
-  // dynamic casting to expose interfaces for these based on type isn't an
-  // option. COM is the usual solution for this. But it's not worth it to add
-  // all of the COM overhead to this class just for that.
-  virtual DataTypeT
-  computeExpectedValue([[maybe_unused]] const DataTypeT &A,
-                       [[maybe_unused]] const DataTypeT &B) const {
-    LOG_ERROR_FMT_THROW(L"E_NOT_IMPL: computeExpectedValue for a Binary Op");
-    return DataTypeT();
-  }
-  virtual DataTypeT
-  computeExpectedValue([[maybe_unused]] const DataTypeT &A) const {
-    LOG_ERROR_FMT_THROW(L"E_NOT_IMPL: computeExpectedValue for a Unary Op");
-    return DataTypeT();
+        Operator(OpTypeMd.Operator),
+        ScalarInputFlags(OpTypeMd.ScalarInputFlags) {}
+
+  // Helper to initialize a unary value computer.
+  template <typename DataTypeOutT>
+  void InitUnaryOpValueComputer(
+      std::function<DataTypeOutT(const DataTypeT &)> ComputeFunc) {
+    DXASSERT_NOMSG(BasicOpType == BasicOpType_Unary);
+    DXASSERT_NOMSG(ExpectedValueComputer == nullptr);
+    ExpectedValueComputer =
+        std::make_unique<UnaryOpExpectedValueComputer<DataTypeT, DataTypeOutT>>(
+            ComputeFunc);
   }
 
+  // Helper to initialize a binary value computer.
+  template <typename DataTypeOutT>
+  void InitBinaryOpValueComputer(
+      std::function<DataTypeOutT(const DataTypeT &, const DataTypeT &)>
+          ComputeFunc) {
+    DXASSERT_NOMSG(BasicOpType == BasicOpType_Binary);
+    DXASSERT_NOMSG(ExpectedValueComputer == nullptr);
+    ExpectedValueComputer = std::make_unique<
+        BinaryOpExpectedValueComputer<DataTypeT, DataTypeOutT>>(ComputeFunc);
+  }
+
+  // Helper to initialize a ternary value computer.
+  template <typename DataTypeOutT>
+  void InitTernaryOpValueComputer(
+      std::function<DataTypeOutT(const DataTypeT &, const DataTypeT &,
+                                 const DataTypeT &)>
+          ComputeFunc) {
+    DXASSERT_NOMSG(BasicOpType == BasicOpType_Ternary);
+    DXASSERT_NOMSG(ExpectedValueComputer == nullptr);
+    ExpectedValueComputer = std::make_unique<
+        TernaryOpExpectedValueComputer<DataTypeT, DataTypeOutT>>(ComputeFunc);
+  }
+
+  std::unique_ptr<ExpectedValueComputerBase<DataTypeT>> ExpectedValueComputer =
+      nullptr;
```

Comment:
Why does this need to be stored? It seems a lot of the complexity I'm trying to understand here is driven from the desire to store and reuse this.

Reply:
It's hard to figure out what the right feedback here is.  I have a gut feeling that there's a simpler design in here somewhere, but nothing that's actually helpful to help find it.  I'm poking at this bit of it because I think it'd be good to have only compile-time or runtime-time polymorphism, but doing both at the same time makes it hard to get my head around.